// lvovich/esm/gender.js
/* eslint-disable no-use-before-define, arrow-parens */
import { startsWith, endsWith } from './utils';
import genderRules from './rules/genderRules';
export const MALE = 1;
export const FEMALE = 2;
export const ANDROGYNOUS = 4;
export function getFG(str) {
    return getGenderByRuleSet(str, genderRules.firstname);
}
export function getLG(str) {
    return getGenderByRuleSet(str, genderRules.lastname);
}
export function getMG(str) {
    return getGenderByRuleSet(str, genderRules.middlename);
}
export function getFirstnameGender(str) {
    return convertGenderStr(getFG(str));
}
export function getLastnameGender(str) {
    return convertGenderStr(getLG(str));
}
export function getMiddlenameGender(str) {
    return convertGenderStr(getMG(str));
}
export function mergeGenders(g1, g2) {
    if (g1 === ANDROGYNOUS)
        return g2;
    if (g2 === ANDROGYNOUS)
        return g1;
    if (g1 === g2)
        return g1;
    return null;
}
export function _getGender(fio) {
    let result = ANDROGYNOUS;
    const { middle, first, last } = fio;
    if (middle) {
        result = mergeGenders(result, getMG(middle.trim()));
    }
    if (first) {
        result = mergeGenders(result, getFG(first.trim()));
    }
    if (last) {
        const lastGender = getLG(last.trim());
        if (lastGender !== null) {
            result = mergeGenders(result, lastGender);
        }
    }
    return result;
}
export function getGender(fio) {
    return convertGenderStr(_getGender(fio));
}
export function getGenderByRuleSet(name, ruleSet) {
    if (!name || !ruleSet) {
        return null;
    }
    const nameLower = name.toLowerCase();
    if (ruleSet.exceptions) {
        const gender = getGenderByRule(ruleSet.exceptions, (some) => {
            if (startsWith(some, '-')) {
                return endsWith(nameLower, some.substr(1));
            }
            return some === nameLower;
        });
        if (gender)
            return gender;
    }
    return ruleSet.suffixes
        ? getGenderByRule(ruleSet.suffixes, (some) => endsWith(nameLower, some))
        : null;
}
export function getGenderByRule(rules, matchFn) {
    const genders = Object.keys(rules).filter((genderKey) => {
        const array = rules[genderKey];
        return Array.isArray(array) && array.some(matchFn);
    });
    if (genders.length !== 1) {
        // DEBUG SEVERAL RULES
        Object.keys(rules).forEach((genderKey) => {
            const array = rules[genderKey];
            if (Array.isArray(array) && array.some(matchFn)) {
                // eslint-disable-next-line
                // console.log(genderKey, array);
            }
        });
        return null;
    }
    return getGenderConst(genders[0]);
}
export function getGenderConst(key) {
    switch (key) {
        case 'male':
        case MALE:
            return MALE;
        case 'female':
        case FEMALE:
            return FEMALE;
        case 'androgynous':
        case ANDROGYNOUS:
            return ANDROGYNOUS;
        default:
            return null;
    }
}
export function convertGenderStr(cnst) {
    switch (cnst) {
        case 'male':
        case MALE:
            return 'male';
        case 'female':
        case FEMALE:
            return 'female';
        case 'androgynous':
        case ANDROGYNOUS:
            return 'androgynous';
        default:
            return null;
    }
}
